### ISSUE TITLE
DateTimeInterface has more methods than listed

### ISSUE THREAD

**Issue by muglug** [vimeo/psalm#2180]
`DateTimeInterface` can only be implemented by `DateTime` and `DateTimeImmutable` (PHP throws an error for userland implementations of `DateTimeImmutable`).

Psalm should not complain about the following:

```php
class Foo
{
    private DateTimeImmutable $t;
  
    public function __construct(DateTimeInterface $t)
    {
        $this->t = $t->setTimezone(new DateTimeZone('UTC'));
    }
}
```

https://psalm.dev/r/f6064ecfb4

**Comment by weirdan** [comment]
That's probably is not very good idea. It is not safe to mutate `DateTime` which you may have received, because caller may rely on it to be unchanged. This, I suppose, is the reason why mutating methods are not declared on `DateTimeInterface`, even though it was technically possible to do.

**Comment by weirdan** [comment]
This is a better reproducer to what was described in the linked PHPStan issue (https://psalm.dev/r/56f3a89b56):
```php
<?php
  
  function safeDate(DateTimeInterface $d): DateTimeImmutable {
     if ($d instanceof DateTime) {
       return DateTimeImmutable::createFromMutable($d);
     }
     return $d;
  }
```
**Expected**: no issues, as `if` + `return` rules out any possibility of `$d` being anything but `DateTimeImmutable` after the `if`
**Actual**: 
```
Psalm output (using commit 9af8a9a):

INFO: LessSpecificReturnStatement - 7:13 - The type 'DateTimeInterface' is more general than the declared return type 'DateTimeImmutable' for safeDate

INFO: MoreSpecificReturnType - 3:44 - The declared return type 'DateTimeImmutable' for safeDate is more specific than the inferred return type 'DateTimeInterface'
```

It's not even a bug, just a missing feature ("Understanding DateTime inheritance" or something along those lines). 

But allowing mutation on undifferentiated `DateTimeInterface` is a big no-no in my opinion.

**Comment by bdsl** [comment]
I would consider the current Psalm behaviour at https://psalm.dev/r/f6064ecfb4 correct. As @weirdan said, the semantics of the functions in the two classes are different, https://www.php.net/manual/en/datetime.settimezone.php https://www.php.net/manual/en/datetimeimmutable.settimezone.php , so it makes sense that this function isn't listed at https://www.php.net/manual/en/class.datetimeinterface.php . That would have to have documentation to cover both cases, not specifying whether or not it mutates the object, which would make it hard to use safely.

I'm not sure if there's any missing feature either. Is there any fundamental reason why there couldn't be new implementations of DateTimeInterface in a future PHP version, either built-in or by allowing userspace classes to implement it? Presumably the PHP core team wouldn't consider creating a new implementation to be a BC break, so Psalm shouldn't assume that one won't exist at runtime.

**Comment by SignpostMarv** [comment]
can PR a fuller implementation of this if it's suitable:

working with https://psalm.dev/r/ea07690003

using changes:
```diff
diff --git a/src/Psalm/Internal/Analyzer/FunctionLikeAnalyzer.php b/src/Psalm/Internal/Analyzer/FunctionLikeAnalyzer.php
index ec27d435..68558afe 100644
--- a/src/Psalm/Internal/Analyzer/FunctionLikeAnalyzer.php
+++ b/src/Psalm/Internal/Analyzer/FunctionLikeAnalyzer.php
@@ -424,6 +424,22 @@ abstract class FunctionLikeAnalyzer extends SourceAnalyzer
             }
         }
 
+        $expect_datetime_interface = [new TNamedObject('DateTimeInterface')];
+
+        foreach ($params as $param) {
+            if ($param->type) {
+                $param_atomics = $param->type->getAtomicTypes();
+
+                if (1 === count($param_atomics)
+                    && isset($param_atomics['DateTimeInterface'])
+                    && $param_atomics['DateTimeInterface'] instanceof TNamedObject
+                    && $param_atomics['DateTimeInterface']->value === 'DateTimeInterface'
+                ) {
+                    $param->type = new Type\Union([new TNamedObject('DateTime'), new TNamedObject('DateTimeImmutable')]);
+                }
+            }
+        }
+
         if ($codebase->alter_code) {
             $this->alterParams($codebase, $storage, $params, $context);
         }
diff --git a/src/Psalm/Internal/Stubs/CoreGenericClasses.php b/src/Psalm/Internal/Stubs/CoreGenericClasses.php
index 3d9b6e22..902bc535 100644
--- a/src/Psalm/Internal/Stubs/CoreGenericClasses.php
+++ b/src/Psalm/Internal/Stubs/CoreGenericClasses.php
@@ -1367,3 +1367,19 @@ final class WeakReference
     /** @return ?T */
     public function get(): ?object;
 }
+
+/** @template T as $this|DateTimeImmutable */
+interface DateTimeInterface
+{
+    /** @return T|false */
+    public function setTimezone(DateTimeZone $timezone);
+}
+
+/** @template-implements DateTimeInterface<$this> */
+class DateTime implements DateTimeInterface {
+}
+
+/** @template-implements DateTimeInterface<DateTimeImmutable> */
+class DateTimeImmutable implements DateTimeInterface {
+    public static function createFromMutable(DateTime $datetime) : DateTimeImmutable;
+}
```
the result is:
> ERROR: PossiblyInvalidPropertyAssignmentValue - foo.php:9:28 - $this->fixedTime with declared type 'DateTimeImmutable' cannot be assigned possibly different type 'DateTime|DateTimeImmutable|false'
>        $this->fixedTime = $fixedTime->setTimezone(new DateTimeZone('UTC'));

**Comment by muglug** [comment]
Going to close and leave as-is
